controller/volume_controller.go

```diff
@@ -1315,29 +1322,61 @@ func (vc *VolumeController) replenishReplicas(v *longhorn.Volume, e *longhorn.En
 		return nil
 	}
 
-	if hardNodeAffinity != "" && getRebuildingReplicaCount(e) == 0 {
-		r, err := vc.createReplica(v, e, rs, hardNodeAffinity)
-		if err != nil {
-			return err
-		}
-		rs[r.Name] = r
+	if currentRebuilding := getRebuildingReplicaCount(e); currentRebuilding != 0 {
+		return nil
 	}
 
+	log := getLoggerForVolume(vc.logger, v)
+
 	replenishCount := vc.getReplenishReplicasCount(v, rs)
-	if len(rs) != 0 && replenishCount > 0 {
-		// For rebuild case, schedule and rebuild one replica at a time
+	// For regular rebuild case or data locality case, rebuild one replica at a time
+	if (len(rs) != 0 && replenishCount > 0) || hardNodeAffinity != "" {
 		replenishCount = 1
-		currentRebuilding := getRebuildingReplicaCount(e)
-		if currentRebuilding != 0 {
-			return nil
-		}
 	}
+
 	for i := 0; i < replenishCount; i++ {
-		r, err := vc.createReplica(v, e, rs, "")
+		hasPotentialReusableReplica, reusableFailedReplica, err := vc.scheduler.CheckAndReuseFailedReplica(rs, v, hardNodeAffinity)
 		if err != nil {
-			return err
+			return errors.Wrapf(err, "failed to reuse a failed replica during replica replenishment")
+		}
+		if reusableFailedReplica != nil {
+			log.Debugf("Failed replica %v will be reused during rebuilding", reusableFailedReplica.Name)
+			reusableFailedReplica.Spec.FailedAt = ""
+			reusableFailedReplica.Spec.HealthyAt = ""
+			reusableFailedReplica.Spec.RebuildRetryCount++
+			rs[reusableFailedReplica.Name] = reusableFailedReplica
+		} else {
+			newReplicaRequired := false
+			// Creating new replica immediately when:
+			// 1. data locality is required
+			// 2. the volume is a new volume
+			// 3. replica eviction happens
+			// 4. there is no potential reusable replica
+			if hardNodeAffinity != "" || v.Status.Robustness != types.VolumeRobustnessDegraded || !hasPotentialReusableReplica {
```

Comment:
Same above. Do we need to check `hasPotentialReusableReplica` since we are already in `reusableFailedReplica == nil` clause?